### PR TITLE
rest api to get network adapters and to get network adapter infor based on dev

### PR DESCRIPTION
**Pull request by bybai**
**rest api to get network adapters, for example** 
]# localrest list adapters
json
["lo        Link encap:Local Loopback  ","eth0      Link encap:Ethernet  HWaddr 42:5F:0A:05:6A:08  "]

**rest api to get network adapter info, for example:**
]# localrest show netres eth0 detail
json
["2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP qlen 1000\n    link/ether 42:5f:0a:05:6a:08 brd ff:ff:ff:ff:ff:ff\n    inet 10.5.106.8/8 brd 10.255.255.255 scope global eth0\n    inet6 fe80::405f:aff:fe05:6a08/64 scope link \n       valid_lft forever preferred_lft forever"]

]#  localrest show netres  detail
json
["Device \"detail\" does not exist."]

### CODE REVIEW

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 1]
```diff
@@ -163,6 +163,36 @@ sub list_adapter {
 
 #-------------------------------------------------------
 
+=head3  handler to list network adapter first ip
+
+  Subroutine to handle rest request
+  GET /localres/netip/<iface>
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: I think netip is not good a resource name. As not a list info, I prefer show_<resource> here.

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 2]
```diff
@@ -163,6 +163,36 @@ sub list_adapter {
 
 #-------------------------------------------------------
 
+=head3  handler to list network adapter first ip
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: Getting first ip has limit the function of this rest api.
If I want all the IP Addresses of this nic?   Then I have to add a new rest api interface to implement my requirement.

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 3]
```diff
@@ -163,6 +163,36 @@ sub list_adapter {
...
+sub list_netip {
+    my @params = @_;
+    my ($rsp, $result, $iface, $cmd);
+    if (!@params) {
+        $rsp->{data}->[0] = "Argmument error.";
+        xCAT::MsgUtils->message("E", $rsp, $::callback);
+        return 1;
+    }
+    $iface = shift @params;
+    $cmd = "ip addr show dev $iface |grep inet|head -1|awk '{print \$2}'|awk -F/ '{print \$1}'";
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: Call the shell script like this is very dangerous.

1. Too much fork, code efficient is affected.
2. You only check whether the `nic` argument is exist above.  How about a wrong nic name given by the rest api ?

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 4]
```diff
@@ -163,6 +163,65 @@ sub list_adapter {
...
+  Subroutine to handle rest request
+  GET /localres/netmask/<iface>
+
+  Usage example:
+        This function is called from handle_rest_request,
+        do not call it directly.
+=cut
+
+#-------------------------------------------------------
+sub list_netmask {
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: Like my comments above, netmask and ip can be looked as a attribute of resource, no need to add new restapi like this.

If your patch is not very long and do not affect the review progress, please realize that too much useless commit will affect the search result of the remote git repo history.

Comment (bybai) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: Hi chenglch, 
I just used this pull requests to keep my tmp code.  I will notify you when code is ready.

Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: @bybai If you just use it for temporary code, you should mention it in the subject of the pull request before wasting reviewer's time.

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 5]
```diff
@@ -156,9 +156,39 @@ sub list_adapter {
         if ($item eq '.' || $item eq '..') {
             next;
         }
-        $result->[ $i++ ] = $item;
+        $cmd = "ifconfig| grep $item";
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: What's the purpose to add this unfriendly command output when listing adpaters?
In addition, I think this command may not available.

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 6]
```diff
@@ -156,9 +156,39 @@ sub list_adapter {
...
+    my @params = @_;
+    my ($rsp, $result, $iface, $cmd);
+    if (!@params) {
+        $rsp->{data}->[0] = "Argmument error.";
+        xCAT::MsgUtils->message("E", $rsp, $::callback);
+        return 1;
+    }
+    $iface = shift @params;
+    $cmd = "ip addr show dev $iface";
+    $result->[ 0 ] = xCAT::Utils->runcmd("$cmd", -1);
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: Use the command output over rest api like this do not meet loose coupling principle.

Comment (bybai) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: hi @chenglch, I want to get all the resource at a time. So just the command output is enough.

File: xCAT-server/lib/xcat/plugins/localrest.pm (hunk 5 above)
Comment (bybai) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: Hi @chenglch , I can get the adapter and its addr, the command output is the exact what I want .

**Comment by bybai** [comment]
Hi @chenglch ,
After I polished code, the new result:
n]# localrest list adapters
json
{"eth0":{"ip":"10.5.106.8/8","mac":"42:5f:0a:05:6a:08"},"eth1":{... ...}}

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 7]
```diff
@@ -153,14 +154,38 @@ sub list_adapter {
...
     $i = 0;
     foreach my $item (@dir) {
-        if ($item eq '.' || $item eq '..') {
+        if ($item eq '.' || $item eq '..' || $item eq 'lo') {
             next;
         }
-        $result->[ $i++ ] = $item;
+        $cmd = "ip addr show dev $item";
+        @cmdres = xCAT::Utils->runcmd("$cmd", -1);
+        if (!@cmdres) {
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: The return value of runcmd subroutine can not determine the failure or success of the command called.

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 8]
```diff
@@ -153,14 +154,38 @@ sub list_adapter {
...
+                    $tmpres->{'ip'} = $ip;
+                }
+            }
+            if ( $in =~ 'ether' ) {
+                @macarr = split(' ',$in);
+                $mac = $macarr[1];
+                $tmpres->{'mac'} = $mac;
+            }
+        }
+        $result->{$item} = $tmpres;
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: Seems no operation to clear $tmpres,  if the `if` condition above can not match, what will happen?

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 9]
```diff
@@ -153,14 +154,38 @@ sub list_adapter {
...
     }
-    return $result;
+    return $result
+
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: please delete the blank line

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 10]
```diff
@@ -134,16 +134,17 @@ sub handle_rest_request {
...
   Usage example:
         This function is called from handle_rest_request,
         do not call it directly.
 =cut
 
 #-------------------------------------------------------
-sub list_adapter {
-    my ($rsp, $result, $i);
+sub list_adapters {
+    my ($rsp, $result, $ifaces, $cmd, $i, $tmpres, @cmdres, $mac, $ip);
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: `$i` variable is declared, but has not been used.

The route cause of this is that json format is invalid.

File: xCAT-server/lib/xcat/plugins/localrest.pm [hunk 11]
```diff
@@ -153,14 +154,38 @@ sub list_adapter {
     closedir(DIR);
     $i = 0;
     foreach my $item (@dir) {
-        if ($item eq '.' || $item eq '..') {
+        if ($item eq '.' || $item eq '..' || $item eq 'lo') {
             next;
         }
-        $result->[ $i++ ] = $item;
+        $cmd = "ip addr show dev $item";
```
Comment (chenglch) [inline xCAT-server/lib/xcat/plugins/localrest.pm]: This command is called within a `for` loop, which will affect the code efficiency.
Could you try `ip addr` command , package a json structure by parsing the output about all the adapters?

**Comment by chenglch** [comment]
Hi @bybai 

    {"eth0":{"ip":"10.5.106.8/8","mac":"42:5f:0a:05:6a:08"},"eth1":{... ...}}

The json format  you given is wrong, as `eth0` or `eth1` can not be looked as a fix key(key/value) field directly at this level.

**Comment by bybai** [comment]
Hi @chenglch ,
The new result：
[root@omc xCAT_plugin]# localrest list adapters
json
{"eth0":[{"ip":"10.5.106.8/8","mac":"42:5f:0a:05:6a:08"}],"eth1":[{"ip":"10.5.106.8/8","mac":"42:5f:0a:05:6a:08"}]}

**Comment by chenglch** [comment]
Hi @bybai ,
Maybe you misunderstand my last comment. The json format is still invalid as `eth0` ,`eth1` are uncertain values which should not be placed at the `key` position.

A json reference for nics information:

```
{
    "nics": [
        {
            "name": "eth0",
            "ip": "10.5.106.8",
            "mac": "42:5f:0a:05:6a:08"
        },
        {
            "name": "eth1",
            "ip": "10.5.106.9",
            "mac": "42:5f:0a:05:6a:09"
        }
    ]
}
```

**Comment by bybai** [comment]
Hi @chenglch , new output is as: [{"ip":"10.5.106.8/8","name":"eth0","mac":"42:5f:0a:05:6a:08"}]

**Comment by chenglch** [comment]
Looks good to me.

**Pull request merged**